### PR TITLE
impl : broadcast multi-messages #97

### LINKED ISSUE #97: [conn] impl broadcastMessages
**Subject**

**Detail**

### PR DESCRIPTION
**Pull request by GunHee5719**
## Related Issue

resolve: #97 

## Description

- implements BroadcastMessages() at conn.go that broadcast multi-messages to all nodes one by one.
- implements test case for BroadcastMessages at `conn_test.go`
- add one more case for TestConnectionPool_BroadcastMessage at `conn_test.go`
- modify setting port at `conn_test.go` and `comm_test.go`

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] Applying goimports
- [x] Test case
- [x] End of Work

### CODE REVIEW

**Comment by GunHee5719** [comment]
I need idea for function naming. I think BroadcastMessages() make confuse with BroadcastMessage()

**Comment by zeroFruit** [comment]
`BroadcastMultiMessages`?

**Comment by GunHee5719** [comment]
Not bad.. But @AgwaB rejected it

**Comment by zeroFruit** [comment]
What's your opinion? @AgwaB

**Comment by zeroFruit** [comment]
or how about declaring 

```
type Broadcaster interface {
    SendMessage
}

type MultiMsgBroadcaster interface {
    SendMessages
}
```
To clarify the responsibility.

ref: https://core.ac.uk/download/pdf/82733544.pdf

**Comment by GunHee5719** [comment]
I think it's good way that divide interface, but also i think SendMessage may make confuse with Send()

**Comment by AgwaB** [comment]
I suggest
```go
type Sharer interface {
       Broadcast()
}

type Distributer interface {
       Broadcast()
}
```

**Comment by GunHee5719** [comment]
Then how about

**Comment by GunHee5719** [comment]
Then how about combining your opinions like this?
```go
type Sharer interface {
      BroadcastSingleMsg()
}

type Distributer interface {
      BroadcastMultipleMsg()
}
```

**Comment by zeroFruit** [comment]
@GunHee5719 @AgwaB 
how about unify interface ?
```
type Broadcaster interface {
    Broadcast(msgList []pb.Message)
}
```
what `Broadcaster` should do is **just sending one message from list to its one connection.**

we don't care whether `msgList`'s message is same or not. just sending one message to one connection.

if messages are same then it is single message broadcasting, if not, multi message broadcasting

File: conn_test.go [hunk 1]
```diff
@@ -195,12 +210,133 @@ func TestConnectionPool_Broadcast(t *testing.T) {
...
+	cp.BroadcastMessage(msg)
 
 	// wait until receive the msg
 	wg.Wait()
 }
 
+func TestConnectionPool_BroadcastMessages(t *testing.T) {
+	cp := cleisthenes.NewConnectionPool()
+
+	cAddress1 := cleisthenes.Address{"127.0.0.1", 8080}
```
Comment (hihiboss) [inline conn_test.go]: Port `8080` is often used. For example, tomcat(web server) uses `8080` as its default port.
So, I recommend to change the port number :)

**Comment by GunHee5719** [comment]
I added code for getting available port and modified the part for setting port at whole `conn_test.go` and `comm_test.go`.

So now I think we need to decide what func name to use

File: conn_test.go [hunk 2]
```diff
@@ -135,29 +151,33 @@ func TestGrpcConnection_Close(t *testing.T) {
...
+		cAddressList[i] = cleisthenes.Address{"127.0.0.1", GetAvailablePort(8000)}
 	}
 
-	mockStreamWrapper2 := mock.NewStreamWrapper()
-	conn2, err2 := cleisthenes.NewConnection(cAddress2, "TestConnection2", mockStreamWrapper2)
-	if err2 != nil {
-		t.Fatal(err2)
+	mockStreamWrapperList := make([]*mock.StreamWrapper, 3)
+	connList := make([]cleisthenes.Connection, 3)
+	var err error
```
Comment (AgwaB) [inline conn_test.go]: what is it?

Comment (GunHee5719) [reply conn_test.go]: I made it because I need error variable for get return value at 168 line.

**Comment by GunHee5719** [comment]
I divided the interface as Sharer & Distributer and pushed. But because of this work, error occurred at RBC struct in `rbc.go` because there are variable `broadCaster cleisthenes.Broadcaster`. Then how can I fix this? Please select options.
 - Make new variable at `rbc.go`
 - Do not divide the interface at `conn.go`
 - etc..

And please give me some advice at naming..

File: conn.go [hunk 3]
```diff
@@ -179,8 +179,14 @@ func (conn *GrpcConnection) readStream(errChan chan error) {
 	}
 }
 
-type Broadcaster interface {
-	Broadcast(msg pb.Message)
+type Sharer interface {
```
Comment (zeroFruit) [inline conn.go]: I think it is better to integrate `Sharer`, `Distributor` into `MessageBroadcaster`

File: conn_test.go [hunk 4]
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
...
 
-	cp.Broadcast(msg)
+	cp.BroadcastSingleMessage(msg)
+
+	// wait until receive the msg
+	wg.Wait()
+}
+
+func Execute_BroadcastMultipleMessage(textList []string, waitSize int) ([]bool, error) {
+	cp := cleisthenes.NewConnectionPool()
```
Comment (zeroFruit) [inline conn_test.go]: ### change function name into camel-case & private function

### refactor large function

I think what this function do is 

1. make connection list
2. register mock handler for each connection
3. ...

I think we can make function prettier & clear

File: conn_test.go [hunk 5]
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
...
+
+	for i, _ := range cAddressList {
+		cAddressList[i] = cleisthenes.Address{"127.0.0.1", GetAvailablePort(8000)}
+	}
+
+	connList := make([]cleisthenes.Connection, 3)
+	var err error
+	err = nil
+
+	for i, _ := range connList {
```
Comment (zeroFruit) [inline conn_test.go]: `for i, _` => `for i`

File: conn_test.go [hunk 6]
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
...
+	}
+
+	for _, conn := range connList {
+		cp.Add(conn.Id(), conn)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(waitSize)
+
+	checkReceive := make([]bool, len(textList))
```
Comment (zeroFruit) [inline conn_test.go]: msgReceivedChecker, msgReceivedCheckList? anyway It would be better variable name ends with *noun*

**Review by zeroFruit — CHANGES_REQUESTED** [review]

**Comment by zeroFruit** [comment]
test failed

**Review by AgwaB — APPROVED** [review]

**Review by zeroFruit — APPROVED** [review]
LGTM

**Pull request merged**